controllers/topology/internal/extensions/patches/engine.go

```diff
@@ -0,0 +1,495 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package patches implement the patch engine.
+package patches
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/pkg/errors"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/contract"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/api"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/variables"
+	tlog "sigs.k8s.io/cluster-api/controllers/topology/internal/log"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/scope"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Apply applies patches to the desired state according to the patches from the ClusterClass, variables from the Cluster
+// and builtin variables.
+// * A GenerateRequest with all templates and global and template-specific variables is created.
+// * Then for all patches of a ClusterClass, patches are generated and successively applied to the templates
```

Comment:
```suggestion
// * Then for all patches definitions of a ClusterClass, patches are generated and successively applied to the templates
```

Reply:
I've went with a slightly different wording to deduplicate "patches"